### PR TITLE
docs: windows installation guidance 

### PR DESCRIPTION
**Pull request by X-jy-liu**
## Description
Briefly describe your changes.
It adds the installation steps for Windows machine

## Motivation
Explain why this change is needed and what problem it solves.
It sums up the solutions to Windows installation in issue `close #123 ` 

## Type of Change
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [x] Documentation update

## Checklist
- [x] I have read the [CONTRIBUTION](https://github.com/ServerlessLLM/ServerlessLLM/blob/main/CONTRIBUTING.md) guide.
- [ ] I have updated the tests (if applicable).
- [ ] I have updated the documentation (if applicable).

### CODE REVIEW

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:56]: Could you help us implement a solution or fix the bug within the code to automatically create this folder without permission issue, rather than requiring manual creation? We want the user to be able to have the folder generated simply by running the code. This issue doesn’t occur on our server, so it may be related to the WSL permission settings. Additionally, it would be preferable to avoid using "I" in the user guide.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:67]: I don't think this issue is related to the WSL. You do not need to include it.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Thank you for your suggestions. I've automated the folder-checking and packed it in the model_folder_checker.py file. I hope this method is okay. I also removed the first person in the .md file.

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Thank you for pointing this out. I've got rid of them.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:23]: Please replace instances of "I" with alternative wording in these texts.

File: examples/installation_on_windows/model_folder_checker.py
Comment (SecretSettler) [inline examples/installation_on_windows/model_folder_checker.py:34]: Hi, we don’t need an additional script for checking and creating folders. Our system already includes a class called `VllmModelDownloader` located in `sllm/serve/model_downloader.py`. The issue seems to be due to the default file permissions set by `shutil.copytree` and `shutil.copy`, which may limit access to the copied folder under `dest_path` in WSL. Please resolve this issue within that file, ensuring the solution doesn’t impact functionality on standard Linux systems.

File: examples/installation_on_windows/cuda-version-check.png
Comment (SecretSettler) [inline examples/installation_on_windows/cuda-version-check.png:1]: Please hide your username.

File: examples/installation_on_windows/wsl-terminal.png
Comment (SecretSettler) [inline examples/installation_on_windows/wsl-terminal.png:1]: Hide the username at the top please.

**Comment by SecretSettler** [comment]
Furthermore, ensure the `CONTRIBUTING.md` file is reviewed for the correct commit, title, and format requirements before submitting the PR. `Pre-commit` checks should be run prior to pushing. A guide is included within the file; please follow it closely.

File: examples/installation_on_windows/model_folder_checker.py
Comment (X-jy-liu) [reply examples/installation_on_windows/model_folder_checker.py:34]: Hi, I was looking into this one today and tested the `shutil` functions that work fine. So I was trying to reproduce the model issue but didn't encounter the access issue again. Everything is downloaded and run smoothly from `pip install`. I was wondering if you updated anything related to this on your side. I'm unsure how this issue was triggered in the first place. It might be caused by something in my machine, so I updated the .md file ignoring this issue for now. But if anyone run into the same issue or if you want me to check something else specifically, I'm more than happy to do that.

File: examples/installation_on_windows/cuda-version-check.png
Comment (X-jy-liu) [reply examples/installation_on_windows/cuda-version-check.png:1]: Thank you for your patience! I've hidden my id from both the images texts.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [reply examples/installation_on_windows/README.md:39]: It seems the spell check has missed some words. Could you please add the words marked as "incorrectly spelled" to `configs/.wordlist.txt` so the auto-checker recognizes them?

File: examples/installation_on_windows/model_folder_checker.py
Comment (SecretSettler) [reply examples/installation_on_windows/model_folder_checker.py]: Okay.

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Hi, thank you again for your reply! I've updated the wsl config part. It mainly depends on my experiments. But I also needed some explanations on how wsl connection manages the memory and swap. I hope these would work.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:38]: Hi,  
Thank you for the modification—it looks great! Could you confirm whether the `sllm-store-server` will fail to run if the WSL settings are configured to less than 22GB of memory and 8GB of swap?

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:56]: Could you please clarify and reorganize the instructions? For running the script, it's preferable to use the following format:  
```bash
cd /path/to/xxx  
python xxx.py  
```  
Actually, the location of `.` for `./models/vllm` can be different from `./examples/installation_on_windows/model_folder_checker.py` if the user did not intially `cd` to the folder `ServerlessLLM`, which creates ambiguity.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md:38]: Hi, I doublechecked it. It turns out that I didn't clear the cache in previous tests. Now I can confirm 17GB Memory and 8GB Swap is the minimum requirements as I updated in the PR. Thank you.

Here is the test result for the configuration, a Memory of 16GB and a Swap of 8GB.
af4597fd401ccda39b71e7c7bfd1929
PS: there was some memory loss while setting the wsl connection.

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Changes have been made accordingly

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:18]: Hi, could you please avoid specifying the current Ubuntu release version? Since we cannot predict the Ubuntu release version users will have when using ServerlessLLM, it would be better to simply provide instructions on how to upgrade.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:6]: It would be preferable to provide another picture in a fully English version.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:4]: Hi, Could you please change SLLM to ServerlessLLM?

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [reply examples/installation_on_windows/README.md:38]: Hi, is this issue related to the `sllm-store-server` not initializing properly, or is it another deployment issue? Were you able to successfully start the `sllm-store-server` with 16GB of memory and 8GB of swap? This error shows a deployment issue, but the store server might be initialized normally.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md:57]: yes, both of them work. Changed accordingly

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md:18]: Thank you for this advice. I removed the detailed descriptions (version number) and left only the key information to reduce ambiguity.

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md:6]: fully English verision updated

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md:4]: Changed accordingly

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: It started this server normally. 4GB memory pool created, server is listen and control loop running, etc. all seem good. But it ran into OOM issue in the deployment. So this should be a deployment issue.

**Comment by SecretSettler** [comment]
Hi, it looks good to me. Please pass your content to ChatGPT for rephrasing, correcting grammar errors, and enhancing coherence.

**Comment by SecretSettler** [comment]
This PR includes a script to check folder permissions, which might occasionally cause issues in the WSL environment. Do you think it is necessary? @future-xy

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Hi,
Could you please change the path to `ServerlessLLM/examples/installation_on_windows/`. Thank you.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:42]: Hi,
Could you please change the path to `ServerlessLLM/examples/installation_on_windows/`. Thank you.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Hi, we have confirmed that this issue is caused by an environment dependency related to the hardware information collection feature we recently released, not an OOM error. Could you please re-test the minimum WSL settings after we resolve this issue? From my perspective, the memory allocation for WSL is tied to the size of the `sllm-store-server` memory pool you have configured. By default, it is set to 4GB. Therefore, the minimum requirement should be less than 16GB Memory and 8GB Swap but not lower than 4GB Mem.

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Hi, changed accordingly

**Comment by X-jy-liu** [comment]
> Hi, it looks good to me. Please pass your content to ChatGPT for rephrasing, correcting grammar errors, and enhancing coherence.

Hi, ChatGPT rephasing done

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: The tests I've been talking so far are based on the version I installed on Friday by
```bash
# On the head node
conda create -n sllm python=3.10 -y
conda activate sllm
pip install serverless-llm
pip install serverless-llm-store

# On a worker node
conda create -n sllm-worker python=3.10 -y
conda activate sllm-worker
pip install serverless-llm[worker]
pip install serverless-llm-store
```

I
I reinstalled the software today, and the testing results are similar to before. However, I noticed that this warning might be the reason behind the unexpected storage requirements. I'm happy to contribute more in future PRs, but I want to make a phased and conclusive contribution for now if that's okay with you guys.

```vbnet
sllm-serve start
2024-11-25 10:52:33,593 WARNING services.py:407 -- Found multiple active Ray instances: {'172.30.30.103:6379', '0.0.0.0:6379'}. Connecting to latest cluster at 0.0.0.0:6379. You can override this by setting the `--address` flag or `RAY_ADDRESS` environment variable.
2024-11-25 10:52:33,593 INFO worker.py:1540 -- Connecting to existing Ray cluster at address: 0.0.0.0:6379...
2024-11-25 10:52:33,598 INFO worker.py:1724 -- Connected to Ray cluster.
(SllmController pid=2625, ip=172.30.30.103) INFO 11-25 10:52:34 controller.py:60] Starting store manager
(SllmController pid=2625) INFO 11-25 10:52:34 controller.py:73] Starting scheduler
(StoreManager pid=2677) INFO 11-25 10:52:34 store_manager.py:196] Initializing store manager
**(StoreManager pid=2677) WARNING 11-25 10:52:34 store_manager.py:208] Hardware info not provided. Storage-aware scheduling might not work properly**
INFO:     Started server process [2399]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8343 (Press CTRL+C to quit)
(FcfsScheduler pid=2729) INFO 11-25 10:52:34 fcfs_scheduler.py:53] Starting FCFS scheduler
(FcfsScheduler pid=2729) INFO 11-25 10:52:34 fcfs_scheduler.py:94] Starting control loop
(SllmController pid=2625, ip=172.30.30.103) INFO 11-25 10:52:38 controller.py:101] Registering new model facebook/opt-1.3b
```

Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Hello,  

We’ve identified that the error you encountered is due to an issue with passing the type parameter to the vLLM model downloader. This issue does not occur when using the Hugging Face Transformers backend. Could you please test the minimum setup with the Transformers backend? You can specify this backend when deploying the model:  

```bash
sllm-cli deploy --model facebook/opt-1.3b --backend transformers
```  

Or you can wait until PR #163 is merged and test again.

The reason we ask you to test the minimum configuration is to ensure users can verify if their machines meet the basic requirements to run our service, even on a simple laptop. This is similar to checking the minimum system requirements provided for games on platforms like Steam (if you’re familiar with that). 

Thank you!

Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: aha, I see. In that case, the minimum configuration needs 4GB Mem available, which makes the minimum setting in the `.wslconfig` be Memory = 5GB and Swap=2.
```bash
sllm-cli deploy --model facebook/opt-1.3b --backend transformers
```
This does improve the progress and makes the downloading process much faster. Also the configuration of 16GB Memory and 8 Swap that failed to deploy the model using `sllm-cli deploy --model facebook/opt-1.3b` can now deploy the model successfully.

Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Okay, thank you. Can you modify the setting?

Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Hi, you can also see there is still an error in spell check, please add these unrecognized words to wordlist.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:56]: Hi, please modify this minimum requirement.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

**Comment by SecretSettler** [comment]
Could you please also modify the PR title. If you are contributing to the examples please add `example: ` at the front.

**Comment by X-jy-liu** [comment]
> Could you please also modify the PR title. If you are contributing to the examples please add `example: ` at the front.

From others PR, I think it should be docs, right?

**Comment by SecretSettler** [comment]
> > Could you please also modify the PR title. If you are contributing to the examples please add `example: ` at the front.
> 
> From others PR, I think it should be docs, right?

That's ok

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:54]: Hi, there’s a typo—it should be opt-1.3b. The dtype issue has been resolved, so please exclude it. Based on your tests, deploying an opt-1.3b model requires 16GB of memory and 8GB of swap, correct? Is memory=5GB, swap=2GB not able to run this model?

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: I was wondering if there is a way to run the github checks locally.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:54]: Hi, there’s a typo—it should be opt-1.3b. The dtype issue has been resolved, so please exclude it. Based on your tests, deploying an opt-1.3b model requires 16GB of memory and 8GB of swap, correct? Is memory=5GB, swap=2GB not able to run this model?

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:63]: Is memory=5GB, swap=2GB not able to run this model?

**Review by SecretSettler — CHANGES_REQUESTED** [review]

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: Yes, it needs 16GB and 8GB swap to run the model. Memory=5GB and Swap = 2GB cannot do it.

Comment (SecretSettler) [reply examples/installation_on_windows/README.md]: Okay, last thing, please exclude the dtype error, as we have resolved this error.

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:54]: Okay, last thing, please exclude the dtype error, as we have resolved this error. Both vLLM and transformers work now.

**Review by SecretSettler — CHANGES_REQUESTED** [review]

**Review by X-jy-liu — COMMENTED** [review]
Please ignore this one. It was a very old commit and pushed by mistake

File: examples/installation_on_windows/README.md
Comment (X-jy-liu) [reply examples/installation_on_windows/README.md]: yes, the resolved error removed

**Review by SecretSettler — APPROVED** [review]
It's good for me. @future-xy Could you please check it?

File: examples/installation_on_windows/README.md
Comment (SecretSettler) [inline examples/installation_on_windows/README.md:104]: What should the user do after this step?

**Review by SecretSettler — CHANGES_REQUESTED** [review]

**Review by SecretSettler — APPROVED** [review]

**Comment by X-jy-liu** [comment]
> Thanks a lot for testing this on Windows and providing this example—great work! 🎉
> 
> I think the document is really useful for beginners who want to try ServerlessLLM on Windows. Would you mind if I make some edits to your branch to align the style with open-source community docs? Let me know what you think. Thanks again for your contribution!

Of course not, @future-xy please go ahead. My pleasure to contribute and looking forward to future ones! Thank for both again for the support and guidance, especially @SecretSettler really appreciate your time and patience. I learned a lot from this experience :)

**Review by future-xy — CHANGES_REQUESTED** [review]
Hi @X-jy-liu and @SecretSettler, 

Could you please take a look at the edit PR I submitted and share your thoughts? Here's the link: [PR#1](https://github.com/X-jy-liu/ServerlessLLM/pull/1).  

Thanks!

**Comment by future-xy** [comment]
Configuring 16GB of memory and 8GB of swap for WSL sounds good to me @X-jy-liu

**Review by future-xy — APPROVED** [review]

**Pull request merged**
